ffpuppet/core.py

```diff
@@ -650,22 +652,52 @@ def create_profile(extension=None, prefs_js=None, template=None):
                 with open(times_json, "w") as times_fp:
                     times_fp.write('{"created":%d}' % (int(time.time()) * 1000))
 
-        # XXX: fuzzpriv extension support
-        # should be removed when bug 1322400 is resolved if it is no longer used
-        if extension is not None:
-            os.mkdir(os.path.join(profile, "extensions"))
-            if os.path.isfile(extension) and extension.endswith(".xpi"):
-                shutil.copyfile(
-                    extension,
-                    os.path.join(profile, "extensions", os.path.basename(extension)))
-            elif os.path.isdir(extension):
-                shutil.copytree(
-                    os.path.abspath(extension),
-                    os.path.join(profile, "extensions", "domfuzz@squarefree.com"))
+        # extension support
+        try:
+            if extension is None:
+                extensions = []
+            elif isinstance(extension, (list, tuple)):
+                extensions = extension
             else:
-                shutil.rmtree(profile, True) # clean up on failure
-                raise RuntimeError("Unknown extension: %r" % extension)
-
+                extensions = [extension]
+            if extensions:
+                os.mkdir(os.path.join(profile, "extensions"))
+            for ext in extensions:
+                if os.path.isfile(ext) and ext.endswith(".xpi"):
+                    shutil.copyfile(
+                        ext,
+                        os.path.join(profile, "extensions", os.path.basename(ext)))
+                elif os.path.isdir(ext):
+                    # read manifest to see what the folder should be named
+                    ext_name = None
+                    if os.path.isfile(os.path.join(ext, "manifest.json")):
+                        try:
+                            with open(os.path.join(ext, "manifest.json")) as manifest:
+                                manifest = json.load(manifest)
+                            ext_name = manifest["applications"]["gecko"]["id"]
+                        except Exception as exc:
+                            log.debug("Failed to parse manifest.json: %s", exc)
+                    elif os.path.isfile(os.path.join(ext, "install.rdf")):
+                        try:
+                            xmlns = {"x": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
+                                     "em": "http://www.mozilla.org/2004/em-rdf#"}
+                            tree = ElementTree.parse(os.path.join(ext, "install.rdf"))
+                            assert tree.getroot().tag == "{%s}RDF" % xmlns["x"]
+                            ids = tree.findall("./x:Description/em:id", namespaces=xmlns)
+                            assert len(ids) == 1
+                            ext_name = ids[0].text
+                        except Exception as exc:
```

Comment:
Another broad except.

Reply:
etree parser raises `xml.etree.ElementTree.ParseError`. I'll update it.